bin/build

```diff
@@ -5,13 +5,34 @@ set -e
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )/.." && pwd )"
 NS=socrata
 BUILD_DIRS=$@
-BUILD_DIRS=${BUILD_DIRS:-`cd $DIR && ls -d */Dockerfile | cut -f1 -d'/'`}
+# If an image name is provided, validate it can be built
+if [ -n "$BUILD_DIRS" ]; then
+  if [ ! -e "$BUILD_DIRS/Dockerfile" ]; then
+    BUILD_DOCKERFILES=$(find ${BUILD_DIRS} -name Dockerfile)
+    rc=$?; if [[ $rc != 0 ]]; then
+      echo "No Dockerfiles found"
+      exit $rc
+    fi
+    # Find all of the available builds for this image name
+    BUILD_DIRS=$(echo "$BUILD_DOCKERFILES" | sed "s:^\(.*\)/Dockerfile:\1:")
+  fi
+fi
+
+BUILD_DIRS=${BUILD_DIRS:-`cd $DIR && find . -name Dockerfile | sed "s:^./\(.*\)/Dockerfile:\1:"`}
 
 function build_image() {
   local subdir=$1
-  local tag=${subdir}
+  local tag=$(echo $subdir | sed "s:/.*$::")
+  local versions=$(echo $subdir | sed -e "s:^${tag}::" -e "s:/: \
+  :g")
 
-  docker build $BUILD_OPTS -t ${NS}/$tag $DIR/$subdir
+  if [ -n "$versions" ]; then
+    for version in $versions; do
+      docker build $BUILD_OPTS -t ${NS}/$tag:$version $DIR/$subdir
```

Comment:
Right.  What we are calling "version" here, in Docker parlance is actually "tag" (and what we're calling "tag" is the image name).  There is an implicit "latest" tag if it is not explicitly set.  But, if an image has already been cached locally, building on top of it won't fetch the latest from the the docker registry.  That would require an explicit `docker pull`.